Close Geth but got panic: sync: WaitGroup is reused before previous Wait has returned

I am running a full node. Today I just restart my geth and got this problem and my state snapshot is broken. I did not forcibly close the geth process, just use `kill geth's pid`  I post the log below

#### System information

Geth version: `1.11.6`
CL client & version: prysm@v4.0.6
OS & Version: Linux
Commit hash : (if `develop`) release version

#### Expected behaviour
Close without panic or at least the data should not broken.

#### Actual behaviour
got panic 'panic: sync: WaitGroup is reused before previous Wait has returned' 


#### Logs:


'INFO [06-19|09:59:48.583] Imported new potential chain segment     number=17,510,679 hash=6e295c..ed8098 blocks=1  txs=93   mgas=7.559   elapsed=63.471ms    mgasps=119.091 dirty=228.54MiB'
'INFO [06-19|09:59:48.626] Chain head was updated                   number=17,510,679 hash=6e295c..ed8098 root=b2514d..c946ad elapsed=1.51769ms'
'INFO [06-19|10:00:00.371] Imported new potential chain segment     number=17,510,680 hash=64359b..3b02ba blocks=1  txs=99   mgas=7.256   elapsed=106.217ms   mgasps=68.311  dirty=228.37MiB'
'INFO [06-19|10:00:00.413] Chain head was updated                   number=17,510,680 hash=64359b..3b02ba root=67e807..962bd7 elapsed=1.546763ms'
'INFO [06-19|10:00:01.371] Got interrupt, shutting down... '
'INFO [06-19|10:00:01.375] HTTP server stopped                      endpoint=192.168.0.142:8545'
'INFO [06-19|10:00:01.375] HTTP server stopped                      endpoint=192.168.0.142:8546'
'INFO [06-19|10:00:02.964] HTTP server stopped                      endpoint=127.0.0.1:8551'
'INFO [06-19|10:00:02.964] IPC endpoint closed                      url=/workdir/execution/geth.ipc'
panic: sync: WaitGroup is reused before previous Wait has returned'
''
'goroutine 415923 [running]:'
'sync.(*WaitGroup).Wait(0xc000618bc0?)'
'        sync/waitgroup.go:118 +0x7f'
'github.com/ethereum/go-ethereum/eth.(*handler).Stop(0xc00fadc000)'
'        github.com/ethereum/go-ethereum/eth/handler.go:575 +0x8b'
'github.com/ethereum/go-ethereum/eth.(*Ethereum).Stop(0xc000620400)'
'        github.com/ethereum/go-ethereum/eth/backend.go:527 +0x50'
'github.com/ethereum/go-ethereum/node.(*Node).stopServices(0xc0001f4b60, {0xc000093a80, 0x1, 0x0?})'
'        github.com/ethereum/go-ethereum/node/node.go:298 +0xbd'
'github.com/ethereum/go-ethereum/node.(*Node).Close(0xc0001f4b60)'
'        github.com/ethereum/go-ethereum/node/node.go:218 +0x18e'
'created by github.com/ethereum/go-ethereum/cmd/utils.StartNode.func1.1'
'        github.com/ethereum/go-ethereum/cmd/utils/cmd.go:92 +0x96'

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -123,7 +123,9 @@ type handler struct {
 
 	chainSync *chainSyncer
 	wg        sync.WaitGroup
-	peerWG    sync.WaitGroup
+
+	activeHandlers atomic.Int32
+	handlerDoneCh  chan struct{}
 }
 
 // newHandler returns a handler for all Ethereum chain management protocol.
@@ -143,6 +145,7 @@ func newHandler(config *handlerConfig) (*handler, error) {
 		merger:         config.Merger,
 		requiredBlocks: config.RequiredBlocks,
 		quitSync:       make(chan struct{}),
+		handlerDoneCh:  make(chan struct{}),
 	}
 	if config.Sync == downloader.FullSync {
 		// The database seems empty as the current block is the genesis. Yet the snap
@@ -288,9 +291,34 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
+// incHandlers increments the number of active handlers if not quitting.
+func (h *handler) incHandlers() bool {
+	if h.chainSync.handlePeerEvent() {
+		h.activeHandlers.Add(1)
+		return true
+	}
+	return false
+}
+
+// decHandlers decrements the number of active handlers, unless quitting, in
+// which case it signals on the handlerDoneCh that it is finished.
+func (h *handler) decHandlers() {
+	select {
+	case <-h.quitSync:
+		h.handlerDoneCh <- struct{}{}
+	default:
+		h.activeHandlers.Add(-1)
+	}
+}
+
 // runEthPeer registers an eth peer into the joint eth/snap peerset, adds it to
 // various subsystems and starts handling messages.
 func (h *handler) runEthPeer(peer *eth.Peer, handler eth.Handler) error {
+	if !h.incHandlers() {
+		return p2p.DiscQuitting
+	}
+	defer h.decHandlers()
+
 	// If the peer has a `snap` extension, wait for it to connect so we can have
 	// a uniform initialization/teardown mechanism
 	snap, err := h.peers.waitSnapExtension(peer)
@@ -298,12 +326,6 @@ func (h *handler) runEthPeer(peer *eth.Peer, handler eth.Handler) error {
 		peer.Log().Error("Snapshot extension barrier failed", "err", err)
 		return err
 	}
-	// TODO(karalabe): Not sure why this is needed
-	if !h.chainSync.handlePeerEvent(peer) {
-		return p2p.DiscQuitting
-	}
-	h.peerWG.Add(1)
-	defer h.peerWG.Done()
 
 	// Execute the Ethereum handshake
 	var (
@@ -359,7 +381,7 @@ func (h *handler) runEthPeer(peer *eth.Peer, handler eth.Handler) error {
 			return err
 		}
 	}
-	h.chainSync.handlePeerEvent(peer)
+	h.chainSync.handlePeerEvent()
 
 	// Propagate existing transactions. new transactions appearing
 	// after this will be sent via broadcasts.
@@ -420,8 +442,10 @@ func (h *handler) runEthPeer(peer *eth.Peer, handler eth.Handler) error {
 // `eth`, all subsystem registrations and lifecycle management will be done by
 // the main `eth` handler to prevent strange races.
 func (h *handler) runSnapExtension(peer *snap.Peer, handler snap.Handler) error {
-	h.peerWG.Add(1)
-	defer h.peerWG.Done()
+	if !h.incHandlers() {
+		return p2p.DiscQuitting
+	}
+	defer h.decHandlers()
 
 	if err := h.peers.registerSnapExtension(peer); err != nil {
 		peer.Log().Warn("Snapshot extension registration failed", "err", err)
@@ -502,7 +526,9 @@ func (h *handler) Stop() {
 	// sessions which are already established but not added to h.peers yet
 	// will exit when they try to register.
 	h.peers.close()
-	h.peerWG.Wait()
+	for i := 0; i < int(h.activeHandlers.Load()); i++ {
+		<-h.handlerDoneCh
+	}
 
 	log.Info("Ethereum protocol stopped")
 }

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -326,11 +326,13 @@ func (h *handler) protoTracker() {
 // incHandlers signals to increment the number of active handlers if not
 // quitting.
 func (h *handler) incHandlers() bool {
-	if h.chainSync.handlePeerEvent() {
+	select {
+	case <-h.quitSync:
+		return false
+	default:
 		h.handlerStartCh <- struct{}{}
 		return true
 	}
-	return false
 }
 
 // decHandlers signals to decrement the number of active handlers.

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -296,10 +296,13 @@ func newHandler(config *handlerConfig) (*handler, error) {
 func (h *handler) protoTracker() {
 	defer h.wg.Done()
 	var (
-		active int
-		done   = make(chan struct{})
+		wg       sync.WaitGroup
+		done     = make(chan struct{})
+		active   int
+		quitting bool
 	)
 	// Start goroutine to track active protocol handlers.
+	wg.Add(1)
 	go func() {
 		for {
 			select {
@@ -308,19 +311,22 @@ func (h *handler) protoTracker() {
 			case <-h.handlerDoneCh:
 				active--
 			case <-done:
+				quitting = true
+			}
+			// Only check if active is 0 after quitSync has been
+			// closed. This ensures that in case active drops to 0
+			// naturally, this tracker routine stays operational.
+			if quitting && active == 0 {
+				wg.Done()
 				return
 			}
 		}
 	}()
 	<-h.quitSync
-
-	// After quit is initiated, poll until all handlers are closed.
-	for range time.Tick(time.Millisecond * 50) {
-		if active == 0 {
-			break
-		}
-	}
 	done <- struct{}{}
+
+	// Wait until all handlers finish.
+	wg.Wait()
 }
 
 // incHandlers signals to increment the number of active handlers if not

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -295,38 +295,21 @@ func newHandler(config *handlerConfig) (*handler, error) {
 // protoTracker tracks the number of active protocol handlers.
 func (h *handler) protoTracker() {
 	defer h.wg.Done()
-	var (
-		wg       sync.WaitGroup
-		done     = make(chan struct{})
-		active   int
-		quitting bool
-	)
-	// Start goroutine to track active protocol handlers.
-	wg.Add(1)
-	go func() {
-		for {
-			select {
-			case <-h.handlerStartCh:
-				active++
-			case <-h.handlerDoneCh:
-				active--
-			case <-done:
-				quitting = true
+	var active int
+	for {
+		select {
+		case <-h.handlerStartCh:
+			active++
+		case <-h.handlerDoneCh:
+			active--
+		case <-h.quitSync:
+			// Wait for all active handlers to finish.
+			for ; active > 0; active-- {
+				<-h.handlerDoneCh
 			}
-			// Only check if active is 0 after quitSync has been
-			// closed. This ensures that in case active drops to 0
-			// naturally, this tracker routine stays operational.
-			if quitting && active == 0 {
-				wg.Done()
-				return
-			}
+			return
 		}
-	}()
-	<-h.quitSync
-	done <- struct{}{}
-
-	// Wait until all handlers finish.
-	wg.Wait()
+	}
 }
 
 // incHandlers signals to increment the number of active handlers if not

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -327,11 +327,10 @@ func (h *handler) protoTracker() {
 // quitting.
 func (h *handler) incHandlers() bool {
 	select {
+	case h.handlerStartCh <- struct{}{}:
+		return true
 	case <-h.quitSync:
 		return false
-	default:
-		h.handlerStartCh <- struct{}{}
-		return true
 	}
 }
 

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -304,9 +304,9 @@ func (h *handler) protoTracker() {
 		for {
 			select {
 			case <-h.handlerStartCh:
-				active--
-			case <-h.handlerDoneCh:
 				active++
+			case <-h.handlerDoneCh:
+				active--
 			case <-done:
 				return
 			}

--- a/eth/handler.go
+++ b/eth/handler.go
@@ -124,7 +124,7 @@ type handler struct {
 	chainSync *chainSyncer
 	wg        sync.WaitGroup
 
-	activeHandlers atomic.Int32
+	handlerStartCh chan struct{}
 	handlerDoneCh  chan struct{}
 }
 
@@ -146,6 +146,7 @@ func newHandler(config *handlerConfig) (*handler, error) {
 		requiredBlocks: config.RequiredBlocks,
 		quitSync:       make(chan struct{}),
 		handlerDoneCh:  make(chan struct{}),
+		handlerStartCh: make(chan struct{}),
 	}
 	if config.Sync == downloader.FullSync {
 		// The database seems empty as the current block is the genesis. Yet the snap
@@ -291,24 +292,50 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
-// incHandlers increments the number of active handlers if not quitting.
+// protoTracker tracks the number of active protocol handlers.
+func (h *handler) protoTracker() {
+	defer h.wg.Done()
+	var (
+		active int
+		done   = make(chan struct{})
+	)
+	// Start goroutine to track active protocol handlers.
+	go func() {
+		for {
+			select {
+			case <-h.handlerStartCh:
+				active--
+			case <-h.handlerDoneCh:
+				active++
+			case <-done:
+				return
+			}
+		}
+	}()
+	<-h.quitSync
+
+	// After quit is initiated, poll until all handlers are closed.
+	for range time.Tick(time.Millisecond * 50) {
+		if active == 0 {
+			break
+		}
+	}
+	done <- struct{}{}
+}
+
+// incHandlers signals to increment the number of active handlers if not
+// quitting.
 func (h *handler) incHandlers() bool {
 	if h.chainSync.handlePeerEvent() {
-		h.activeHandlers.Add(1)
+		h.handlerStartCh <- struct{}{}
 		return true
 	}
 	return false
 }
 
-// decHandlers decrements the number of active handlers, unless quitting, in
-// which case it signals on the handlerDoneCh that it is finished.
+// decHandlers signals to decrement the number of active handlers.
 func (h *handler) decHandlers() {
-	select {
-	case <-h.quitSync:
-		h.handlerDoneCh <- struct{}{}
-	default:
-		h.activeHandlers.Add(-1)
-	}
+	h.handlerDoneCh <- struct{}{}
 }
 
 // runEthPeer registers an eth peer into the joint eth/snap peerset, adds it to
@@ -510,6 +537,10 @@ func (h *handler) Start(maxPeers int) {
 	// start sync handlers
 	h.wg.Add(1)
 	go h.chainSync.loop()
+
+	// start peer handler tracker
+	h.wg.Add(1)
+	go h.protoTracker()
 }
 
 func (h *handler) Stop() {
@@ -519,16 +550,13 @@ func (h *handler) Stop() {
 	// Quit chainSync and txsync64.
 	// After this is done, no new peers will be accepted.
 	close(h.quitSync)
-	h.wg.Wait()
 
 	// Disconnect existing sessions.
 	// This also closes the gate for any new registrations on the peer set.
 	// sessions which are already established but not added to h.peers yet
 	// will exit when they try to register.
 	h.peers.close()
-	for i := 0; i < int(h.activeHandlers.Load()); i++ {
-		<-h.handlerDoneCh
-	}
+	h.wg.Wait()
 
 	log.Info("Ethereum protocol stopped")
 }

--- a/eth/handler_eth.go
+++ b/eth/handler_eth.go
@@ -134,7 +134,7 @@ func (h *ethHandler) handleBlockBroadcast(peer *eth.Peer, block *types.Block, td
 	// Update the peer's total difficulty if better than the previous
 	if _, td := peer.Head(); trueTD.Cmp(td) > 0 {
 		peer.SetHead(trueHead, trueTD)
-		h.chainSync.handlePeerEvent(peer)
+		h.chainSync.handlePeerEvent()
 	}
 	return nil
 }

--- a/eth/sync.go
+++ b/eth/sync.go
@@ -89,7 +89,7 @@ func newChainSyncer(handler *handler) *chainSyncer {
 // handlePeerEvent notifies the syncer about a change in the peer set.
 // This is called for new peers and every time a peer announces a new
 // chain head.
-func (cs *chainSyncer) handlePeerEvent(peer *eth.Peer) bool {
+func (cs *chainSyncer) handlePeerEvent() bool {
 	select {
 	case cs.peerEventCh <- struct{}{}:
 		return true

